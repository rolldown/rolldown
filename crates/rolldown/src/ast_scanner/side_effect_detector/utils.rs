use oxc::{
  ast::{
    ast::{Expression, MemberExpression},
    comments_range, Comment, CommentKind,
  },
  semantic::ReferenceId,
  span::{Atom, Span},
  syntax::operator::{BinaryOperator, LogicalOperator, UnaryOperator, UpdateOperator},
};
use rolldown_common::AstScopes;
use rolldown_ecmascript::ExpressionExt;

use super::SideEffectDetector;

impl<'a> SideEffectDetector<'a> {
  /// Get the nearest comment before the `span`, return `None` if no leading comment is founded.
  ///
  ///  # Examples
  /// ```javascript
  /// /* valid comment for `a`  */ let a = 1;
  ///
  /// // valid comment for `b`
  /// let b = 1;
  ///
  /// // valid comment for `c`
  ///
  ///
  /// let c = 1;
  ///
  /// let d = 1; /* valid comment for `e` */
  /// let e = 2
  /// ```
  /// Derived from https://github.com/oxc-project/oxc/blob/147864cfeb112df526bb83d5b8671b465c005066/crates/oxc_linter/src/utils/tree_shaking.rs#L204
  pub fn leading_comment_for(&self, span: Span) -> Option<(&Comment, &str)> {
    let comment = comments_range(self.comments, ..span.start).next_back()?;

    let comment_text = comment.span.source_text(self.source);
    // If there are non-whitespace characters between the `comment` and the `span`,
    // we treat the `comment` not belongs to the `span`.
    let range_text = Span::new(comment.span.end, span.start).source_text(self.source);
    let only_whitespace = match comment.kind {
      CommentKind::Line => range_text.trim().is_empty(),
      CommentKind::Block => {
        range_text
          .strip_prefix("*/") // for multi-line comment
          .is_some_and(|s| s.trim().is_empty())
      }
    };
    if !only_whitespace {
      return None;
    }

    Some((comment, comment_text))
  }
}

#[derive(PartialEq, Eq, Copy, Clone)]
pub(crate) enum PrimitiveType {
  Null,
  Undefined,
  Boolean,
  Number,
  String,
  BigInt,
  Mixed,
  Unknown,
}

fn merged_known_primitive_types(
  scope: &AstScopes,
  left: &Expression,
  right: &Expression,
) -> PrimitiveType {
  let left_type = known_primitive_type(scope, left);
  if left_type == PrimitiveType::Unknown {
    return PrimitiveType::Unknown;
  }
  let right_type = known_primitive_type(scope, right);
  if right_type == PrimitiveType::Unknown {
    return PrimitiveType::Unknown;
  }
  if right_type == left_type {
    return right_type;
  }
  PrimitiveType::Mixed
}

#[allow(clippy::too_many_lines)]
pub(crate) fn known_primitive_type(scope: &AstScopes, expr: &Expression) -> PrimitiveType {
  match expr {
    Expression::NullLiteral(_) => PrimitiveType::Null,
    Expression::Identifier(id)
      if id.name == "undefined" && scope.is_unresolved(id.reference_id.get().unwrap()) =>
    {
      PrimitiveType::Undefined
    }
    Expression::BooleanLiteral(_) => PrimitiveType::Boolean,
    Expression::NumericLiteral(_) => PrimitiveType::Number,
    Expression::StringLiteral(_) => PrimitiveType::String,
    Expression::BigIntLiteral(_) => PrimitiveType::BigInt,
    Expression::TemplateLiteral(e) => {
      if e.expressions.is_empty() {
        PrimitiveType::String
      } else {
        PrimitiveType::Unknown
      }
    }
    Expression::UpdateExpression(e) => {
      match e.operator {
        UpdateOperator::Increment | UpdateOperator::Decrement => {
          PrimitiveType::Mixed // Can be number or bigint
        }
      }
    }
    Expression::UnaryExpression(e) => match e.operator {
      UnaryOperator::Void => PrimitiveType::Undefined,
      UnaryOperator::Typeof => PrimitiveType::String,
      UnaryOperator::LogicalNot | UnaryOperator::Delete => PrimitiveType::Boolean,
      UnaryOperator::UnaryPlus => PrimitiveType::Number, // Cannot be bigint because that throws an exception
      UnaryOperator::UnaryNegation | UnaryOperator::BitwiseNot => {
        let value = known_primitive_type(scope, &e.argument);
        if value == PrimitiveType::BigInt {
          return PrimitiveType::BigInt;
        }
        if value != PrimitiveType::Unknown && value != PrimitiveType::Mixed {
          return PrimitiveType::Number;
        }
        PrimitiveType::Mixed // Can be number or bigint
      }
    },
    Expression::LogicalExpression(e) => match e.operator {
      LogicalOperator::Or | LogicalOperator::And => {
        merged_known_primitive_types(scope, &e.left, &e.right)
      }
      LogicalOperator::Coalesce => {
        let left = known_primitive_type(scope, &e.left);
        let right = known_primitive_type(scope, &e.right);
        if left == PrimitiveType::Null || left == PrimitiveType::Undefined {
          return right;
        }
        if left != PrimitiveType::Unknown {
          if left != PrimitiveType::Mixed {
            return left; // Definitely not null or undefined
          }
          if right != PrimitiveType::Unknown {
            return PrimitiveType::Mixed; // Definitely some kind of primitive
          }
        }
        PrimitiveType::Unknown
      }
    },
    Expression::BinaryExpression(e) => match e.operator {
      BinaryOperator::StrictEquality
      | BinaryOperator::StrictInequality
      | BinaryOperator::Equality
      | BinaryOperator::Inequality
      | BinaryOperator::LessThan
      | BinaryOperator::GreaterThan
      | BinaryOperator::GreaterEqualThan
      | BinaryOperator::LessEqualThan
      | BinaryOperator::Instanceof
      | BinaryOperator::In => PrimitiveType::Boolean,
      BinaryOperator::Addition => {
        let left = known_primitive_type(scope, &e.left);
        let right = known_primitive_type(scope, &e.right);
        if left == PrimitiveType::String || right == PrimitiveType::String {
          PrimitiveType::String
        } else if left == PrimitiveType::BigInt && right == PrimitiveType::BigInt {
          PrimitiveType::BigInt
        } else if !matches!(
          left,
          PrimitiveType::Unknown | PrimitiveType::Mixed | PrimitiveType::BigInt
        ) && !matches!(
          right,
          PrimitiveType::Unknown | PrimitiveType::Mixed | PrimitiveType::BigInt
        ) {
          PrimitiveType::Number
        } else {
          PrimitiveType::Mixed // Can be number or bigint or string (or an exception)
        }
      }
      BinaryOperator::Subtraction
      | BinaryOperator::Multiplication
      | BinaryOperator::Division
      | BinaryOperator::Remainder
      | BinaryOperator::Exponential
      | BinaryOperator::BitwiseAnd
      | BinaryOperator::BitwiseOR
      | BinaryOperator::ShiftRight
      | BinaryOperator::ShiftLeft
      | BinaryOperator::ShiftRightZeroFill
      | BinaryOperator::BitwiseXOR => PrimitiveType::Mixed,
    },

    Expression::AssignmentExpression(e) => match e.operator {
      oxc::syntax::operator::AssignmentOperator::Assign => known_primitive_type(scope, &e.right),
      oxc::syntax::operator::AssignmentOperator::Addition => {
        let right = known_primitive_type(scope, &e.right);
        if right == PrimitiveType::String {
          PrimitiveType::String
        } else {
          PrimitiveType::Mixed // Can be number or bigint or string (or an exception)
        }
      }
      oxc::syntax::operator::AssignmentOperator::Subtraction
      | oxc::syntax::operator::AssignmentOperator::Multiplication
      | oxc::syntax::operator::AssignmentOperator::Division
      | oxc::syntax::operator::AssignmentOperator::Remainder
      | oxc::syntax::operator::AssignmentOperator::ShiftLeft
      | oxc::syntax::operator::AssignmentOperator::ShiftRight
      | oxc::syntax::operator::AssignmentOperator::ShiftRightZeroFill
      | oxc::syntax::operator::AssignmentOperator::BitwiseOR
      | oxc::syntax::operator::AssignmentOperator::BitwiseXOR
      | oxc::syntax::operator::AssignmentOperator::BitwiseAnd
      | oxc::syntax::operator::AssignmentOperator::LogicalAnd
      | oxc::syntax::operator::AssignmentOperator::LogicalOr
      | oxc::syntax::operator::AssignmentOperator::LogicalNullish
      | oxc::syntax::operator::AssignmentOperator::Exponential => PrimitiveType::Mixed,
    },
    _ => PrimitiveType::Unknown,
  }
}

pub fn can_change_strict_to_loose(scope: &AstScopes, a: &Expression, b: &Expression) -> bool {
  let x = known_primitive_type(scope, a);
  let y = known_primitive_type(scope, b);
  x == y && !matches!(x, PrimitiveType::Unknown | PrimitiveType::Mixed)
}

pub fn is_primitive_literal(scope: &AstScopes, expr: &Expression) -> bool {
  match expr {
    Expression::NullLiteral(_)
    | Expression::BooleanLiteral(_)
    | Expression::NumericLiteral(_)
    | Expression::StringLiteral(_)
    | Expression::BigIntLiteral(_) => true,
    // Include `+1` / `-1`.
    Expression::UnaryExpression(e)
      if matches!(e.operator, |UnaryOperator::UnaryNegation| UnaryOperator::UnaryPlus)
        && matches!(e.argument, Expression::NumericLiteral(_)) =>
    {
      true
    }
    Expression::Identifier(id)
      if id.name == "undefined" && scope.is_unresolved(id.reference_id.get().unwrap()) =>
    {
      true
    }
    _ => false,
  }
}

pub fn extract_member_expr_chain<'a>(
  expr: &'a MemberExpression,
  max_len: usize,
) -> Option<(ReferenceId, Vec<Atom<'a>>)> {
  if max_len == 0 {
    return None;
  }
  let mut chain = vec![];
  match expr {
    MemberExpression::ComputedMemberExpression(computed_expr) => {
      let Expression::StringLiteral(ref str) = computed_expr.expression else {
        return None;
      };
      chain.push(str.value.clone());
      let mut cur = &computed_expr.object;
      extract_rest_member_expr_chain(&mut cur, &mut chain, max_len).map(|ref_id| (ref_id, chain))
    }
    MemberExpression::StaticMemberExpression(static_expr) => {
      let mut cur = &static_expr.object;
      chain.push(static_expr.property.name.clone());
      extract_rest_member_expr_chain(&mut cur, &mut chain, max_len).map(|ref_id| (ref_id, chain))
    }
    MemberExpression::PrivateFieldExpression(_) => None,
  }
}

fn extract_rest_member_expr_chain<'a>(
  cur: &mut &'a Expression,
  chain: &mut Vec<Atom<'a>>,
  max_len: usize,
) -> Option<ReferenceId> {
  loop {
    match &cur {
      Expression::StaticMemberExpression(expr) => {
        *cur = &expr.object;
        chain.push(expr.property.name.clone());
      }
      Expression::ComputedMemberExpression(expr) => {
        let Expression::StringLiteral(ref str) = expr.expression else {
          break;
        };
        chain.push(str.value.clone());
        *cur = &expr.object;
      }
      Expression::Identifier(ident) => {
        chain.push(ident.name.clone());
        let ref_id = ident.reference_id.get().expect("should have reference_id");
        chain.reverse();
        return Some(ref_id);
      }
      _ => break,
    }
    // If chain exceeds the max length, that means we are not interest in this member expression.
    // return `None`
    if chain.len() >= max_len {
      return None;
    }
  }
  None
}

/// https://github.com/evanw/esbuild/blob/d34e79e2a998c21bb71d57b92b0017ca11756912/internal/js_ast/js_ast_helpers.go#L2594-L2639
pub fn is_side_effect_free_unbound_identifier_ref(
  scope: &AstScopes,
  value: &Expression,
  guard_condition: &Expression,
  mut is_yes_branch: bool,
) -> Option<bool> {
  let ident = value.as_identifier()?;
  let is_unresolved = scope.is_unresolved(ident.reference_id()?);
  if !is_unresolved {
    return Some(false);
  }
  let bin_expr = guard_condition.as_binary_expression()?;
  match bin_expr.operator {
    BinaryOperator::StrictEquality
    | BinaryOperator::StrictInequality
    | BinaryOperator::Equality
    | BinaryOperator::Inequality => {
      let (mut ty_of, mut string) = (&bin_expr.left, &bin_expr.right);
      if matches!(ty_of, Expression::StringLiteral(_)) {
        std::mem::swap(&mut string, &mut ty_of);
      }
      let unary = ty_of.as_unary_expression()?;
      if !(unary.operator == UnaryOperator::Typeof
        && matches!(unary.argument, Expression::Identifier(_)))
      {
        return Some(false);
      }
      let string = string.as_string_literal()?;

      if (string.value.eq("undefined") == is_yes_branch)
        == matches!(
          bin_expr.operator,
          BinaryOperator::Inequality | BinaryOperator::StrictInequality
        )
      {
        let type_of_value = unary.argument.as_identifier()?;
        if type_of_value.name == ident.name {
          return Some(true);
        }
      }
    }
    BinaryOperator::LessThan
    | BinaryOperator::LessEqualThan
    | BinaryOperator::GreaterThan
    | BinaryOperator::GreaterEqualThan => {
      let (mut ty_of, mut string) = (&bin_expr.left, &bin_expr.right);
      if matches!(ty_of, Expression::StringLiteral(_)) {
        std::mem::swap(&mut string, &mut ty_of);
        is_yes_branch = !is_yes_branch;
      }

      let unary = ty_of.as_unary_expression()?;
      if !(unary.operator == UnaryOperator::Typeof
        && matches!(unary.argument, Expression::Identifier(_)))
      {
        return Some(false);
      }

      let string = string.as_string_literal()?;

      if string.value == "u"
        && is_yes_branch
          == matches!(bin_expr.operator, BinaryOperator::LessThan | BinaryOperator::LessEqualThan)
      {
        let type_of_value = unary.argument.as_identifier()?;
        if type_of_value.name == ident.name {
          return Some(true);
        }
      }
    }
    _ => {}
  }
  Some(false)
}
