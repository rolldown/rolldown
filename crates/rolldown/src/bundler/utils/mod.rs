pub mod ast_scope;
pub mod bitset;
pub mod renamer;
pub mod resolve_id;
