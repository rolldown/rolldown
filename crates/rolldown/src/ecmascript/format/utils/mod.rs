use itertools::Itertools;
use rolldown_std_utils::OptionExt;

use crate::{
  types::generator::GenerateContext,
  utils::chunk::collect_render_chunk_imports::{
    collect_render_chunk_imports, ExternalRenderImportStmt, RenderImportDeclarationSpecifier,
    RenderImportStmt,
  },
};

pub mod namespace;

pub fn render_factory_parameters(
  ctx: &mut GenerateContext<'_>,
  externals: &[ExternalRenderImportStmt],
  has_exports: bool,
) -> String {
  let mut parameters = if has_exports { vec!["exports"] } else { vec![] };
  externals.iter().for_each(|external| {
    let symbol_name = ctx.chunk.canonical_name_by_token.get(&external.binding_name_token).unpack();
    parameters.push(symbol_name.as_str());
  });
  parameters.join(", ")
}

pub fn render_chunk_external_imports(
  ctx: &GenerateContext<'_>,
) -> (String, Vec<ExternalRenderImportStmt>) {
  let render_import_stmts =
    collect_render_chunk_imports(ctx.chunk, ctx.link_output, ctx.chunk_graph);

  let mut import_code = String::new();
  let externals = render_import_stmts
    .into_iter()
    .filter_map(|stmt| {
      if let RenderImportStmt::ExternalRenderImportStmt(external_stmt) = stmt {
        let symbol_name =
          ctx.chunk.canonical_name_by_token.get(&external_stmt.binding_name_token).unpack();
        match &external_stmt.specifiers {
          RenderImportDeclarationSpecifier::ImportSpecifier(specifiers) => {
            // Empty specifiers can be ignored in IIFE.
            if specifiers.is_empty() {
              None
            } else {
              let specifiers = specifiers
                .iter()
                .map(|specifier| {
                  if let Some(alias) = &specifier.alias {
                    format!("{}: {alias}", specifier.imported)
                  } else {
                    specifier.imported.to_string()
                  }
                })
                .collect::<Vec<_>>();
              import_code
                .push_str(&format!("const {{ {} }} = {symbol_name};\n", specifiers.join(", ")));
              Some(external_stmt)
            }
          }
          RenderImportDeclarationSpecifier::ImportStarSpecifier(alias) => {
            import_code.push_str(&format!("const {alias} = {symbol_name};\n"));
            Some(external_stmt)
          }
        }
      } else {
        None
      }
    })
    .collect_vec();

  (import_code, externals)
}
