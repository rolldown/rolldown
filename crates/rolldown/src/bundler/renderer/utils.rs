use std::borrow::Cow;

use oxc::span::{Atom, Span};
use rolldown_common::SymbolRef;
use string_wizard::UpdateOptions;

use crate::bundler::{
  linker::linker_info::LinkingInfo,
  module::{Module, NormalModule},
};

use super::{AstRenderContext, AstRenderer};
impl<'r> AstRenderContext<'r> {
  pub fn canonical_name_for(&self, symbol: SymbolRef) -> &'r Atom {
    self.graph.symbols.canonical_name_for(symbol, self.canonical_names)
  }

  pub fn need_to_rename(&self, symbol: SymbolRef) -> Option<&Atom> {
    let canonical_ref = self.graph.symbols.par_canonical_ref_for(symbol);
    self.canonical_names.get(&canonical_ref)
  }

  pub fn canonical_name_for_runtime(&self, name: &str) -> &Atom {
    let symbol = self.graph.runtime.resolve_symbol(name);
    self.canonical_name_for(symbol)
  }

  pub fn importee_by_span(&self, span: Span) -> &Module {
    &self.graph.modules[self.module.importee_id_by_span(span)]
  }

  pub fn remove_node(&mut self, span: Span) {
    self.source.remove(span.start, span.end);
  }

  pub fn hoisted_module_declaration(&mut self, decl_start: u32, content: String) {
    let start = self.first_stmt_start.unwrap_or(decl_start);
    self.source.append_left(start, content);
  }

  pub fn generate_import_commonjs_module(
    &self,
    importee: &NormalModule,
    importee_linking_info: &LinkingInfo,
    with_declaration: bool,
  ) -> String {
    let wrap_ref_name = self.canonical_name_for(importee_linking_info.wrapper_ref.unwrap());
    let to_esm_ref_name = self.canonical_name_for_runtime("__toESM");
    let code = format!(
      "{to_esm_ref_name}({wrap_ref_name}(){})",
      if self.module.module_type.is_esm() { ", 1" } else { "" }
    );
    if with_declaration {
      let symbol_ref =
        self.linking_info.local_symbol_for_import_cjs.get(&importee.id).copied().unwrap_or_else(
          || {
            panic!(
              "Cannot find local symbol for importee: {:?} with importer {:?} {:?}",
              importee.resource_id, self.module.resource_id, self.module.exports_kind
            )
          },
        );
      let final_name = self.canonical_name_for(symbol_ref);
      format!("var {final_name} = {code};\n")
    } else {
      code
    }
  }
}

impl<'r> AstRenderer<'r> {
  pub fn is_global_require(&self, callee: &oxc::ast::ast::Expression<'_>) -> bool {
    matches!(callee, oxc::ast::ast::Expression::Identifier(ident) if
      ident.name == "require"
      && self.ctx.module.scope.is_unresolved(ident.reference_id.get().unwrap())
    )
  }

  pub fn overwrite(&mut self, start: u32, end: u32, content: String) {
    self.ctx.source.update_with(
      start,
      end,
      content,
      UpdateOptions { overwrite: true, ..Default::default() },
    );
  }

  pub fn generate_namespace_variable_declaration(&mut self) -> Option<String> {
    if self.ctx.module.is_namespace_referenced() {
      let ns_name = self.ctx.canonical_name_for(self.ctx.module.namespace_symbol);
      let exports: String = self
        .ctx
        .linking_info
        .sorted_exports()
        .map(|(exported_name, resolved_export)| {
          let canonical_ref =
            self.ctx.graph.symbols.par_canonical_ref_for(resolved_export.symbol_ref);
          let symbol = self.ctx.graph.symbols.get(canonical_ref);
          let return_expr = if let Some(ns_alias) = &symbol.namespace_alias {
            let canonical_ns_name = &self.ctx.canonical_names[&ns_alias.namespace_ref];
            format!("{canonical_ns_name}.{exported_name}",)
          } else {
            let canonical_name = self.ctx.canonical_names.get(&canonical_ref).unwrap();
            format!("{canonical_name}",)
          };
          format!("{}get {exported_name}() {{ return {return_expr} }}", self.indentor)
        })
        .collect::<Vec<_>>()
        .join(",\n");
      Some(format!("var {ns_name} = {{\n{exports}\n}};\n",))
    } else {
      None
    }
  }

  pub fn remove_node(&mut self, span: Span) {
    self.ctx.source.remove(span.start, span.end);
  }

  pub fn rewrite_symbol(
    &mut self,
    symbol_ref: SymbolRef,
    original_name: &Atom,
    pos: Span,
    is_callee: bool,
  ) {
    let canonical_ref = self.ctx.graph.symbols.par_canonical_ref_for(symbol_ref);
    let symbol = self.ctx.graph.symbols.get(canonical_ref);
    let rendered_symbol = if let Some(ns_alias) = &symbol.namespace_alias {
      // If import symbol from commonjs, the reference of the symbol is not resolved,
      // Here need write it to property access. eg `import { a } from 'cjs'; console.log(a)` => `console.log(import_a.a)`
      // Note: we should rewrite call expression to indirect call, eg `import { a } from 'cjs'; console.log(a())` => `console.log((0, import_a.a)())`
      let canonical_ns_name = self.canonical_name_for(ns_alias.namespace_ref);
      let content = format!("{canonical_ns_name}.{}", ns_alias.property_name);
      Cow::Owned(content)
    } else if let Some(name) = self.need_to_rename(symbol_ref) {
      Cow::Borrowed(name.as_str())
    } else {
      return;
    };

    if original_name != &rendered_symbol {
      self.ctx.source.update(
        pos.start,
        pos.end,
        if is_callee { format!("(0, {rendered_symbol})",) } else { rendered_symbol.into_owned() },
      );
    }
  }

  pub fn canonical_name_for(&self, symbol: SymbolRef) -> &'r Atom {
    self.ctx.graph.symbols.canonical_name_for(symbol, self.ctx.canonical_names)
  }

  pub fn canonical_name_for_runtime(&self, name: &str) -> &Atom {
    let symbol = self.ctx.graph.runtime.resolve_symbol(&Atom::new_inline(name));
    self.canonical_name_for(symbol)
  }

  pub fn need_to_rename(&self, symbol: SymbolRef) -> Option<&Atom> {
    let canonical_ref = self.ctx.graph.symbols.par_canonical_ref_for(symbol);
    self.ctx.canonical_names.get(&canonical_ref)
  }

  pub fn hoisted_module_declaration(&mut self, decl_start: u32, content: String) {
    let start = self.ctx.first_stmt_start.unwrap_or(decl_start);
    self.ctx.source.append_left(start, content);
  }
}
