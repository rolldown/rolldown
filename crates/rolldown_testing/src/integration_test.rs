use std::{borrow::Cow, path::Path, process::Command};

use anyhow::Context;
use rolldown::{
  plugin::__inner::SharedPluginable, BundleOutput, Bundler, BundlerOptions, IsExternal,
  OutputFormat, SourceMapType,
};
use rolldown_common::Output;
use rolldown_error::DiagnosticOptions;
use rolldown_sourcemap::SourcemapVisualizer;
use rolldown_testing_config::TestMeta;
use sugar_path::SugarPath;

use crate::utils::RUNTIME_MODULE_OUTPUT_RE;

#[derive(Default)]
pub struct IntegrationTest {
  test_meta: TestMeta,
}

fn default_test_input_item() -> rolldown::InputItem {
  rolldown::InputItem { name: Some("main".to_string()), import: "./main.js".to_string() }
}

impl IntegrationTest {
  pub fn new(test_meta: TestMeta) -> Self {
    Self { test_meta }
  }

  pub async fn bundle(&self, mut options: BundlerOptions) -> BundleOutput {
    self.apply_test_defaults(&mut options);

    let mut bundler = Bundler::new(options);

    if self.test_meta.write_to_disk {
      if bundler.options().dir.as_path().is_dir() {
        std::fs::remove_dir_all(&bundler.options().dir)
          .context(bundler.options().dir.clone())
          .expect("Failed to clean the output directory");
      }
      bundler.write().await.unwrap()
    } else {
      bundler.generate().await.unwrap()
    }
  }

  pub async fn run(&self, options: BundlerOptions) {
    self.run_with_plugins(options, vec![]).await;
  }

  pub async fn run_with_plugins(
    &self,
    mut options: BundlerOptions,
    plugins: Vec<SharedPluginable>,
  ) {
    self.apply_test_defaults(&mut options);

    let mut bundler = Bundler::with_plugins(options, plugins);

    let cwd = bundler.options().cwd.clone();

    let bundle_output = if self.test_meta.write_to_disk {
      let abs_output_dir = cwd.join(&bundler.options().dir);
      if abs_output_dir.is_dir() {
        std::fs::remove_dir_all(&abs_output_dir)
          .context(format!("{abs_output_dir:?}"))
          .expect("Failed to clean the output directory");
      }
      bundler.write().await.unwrap()
    } else {
      bundler.generate().await.unwrap()
    };

    assert!(
      !(self.test_meta.expect_error && bundle_output.errors.is_empty()),
      "Expected the bundling to be failed with diagnosable errors, but got success"
    );

    assert!(
      !(!bundle_output.errors.is_empty() && !self.test_meta.expect_error),
      "Expected the bundling to be success, but got diagnosable errors: {:?}",
      bundle_output.errors
    );

    self.snapshot_bundle_output(bundle_output, &cwd);

    if !self.test_meta.expect_executed
      || self.test_meta.expect_error
      || !self.test_meta.write_to_disk
    {
      // do nothing
    } else {
      Self::execute_output_assets(&bundler);
    }
  }

  fn apply_test_defaults(&self, options: &mut BundlerOptions) {
    if options.external.is_none() {
      options.external = Some(IsExternal::from_vec(vec!["node:assert".to_string()]));
    }

    if options.input.is_none() {
      options.input = Some(vec![default_test_input_item()]);
    }

    // if options.cwd.is_none() {
    //   options.cwd = Some(fixture_path.to_path_buf());
    // }

    let output_ext = match options.format {
      Some(OutputFormat::Cjs) => "cjs",
      _ => "mjs",
    };

    if options.entry_filenames.is_none() {
      if self.test_meta.hash_in_filename {
        options.entry_filenames = Some(format!("[name]-[hash].{output_ext}"));
      } else {
        options.entry_filenames = Some(format!("[name].{output_ext}"));
      }
    }

    if options.chunk_filenames.is_none() {
      if self.test_meta.hash_in_filename {
        options.chunk_filenames = Some(format!("[name]-[hash].{output_ext}"));
      } else {
        options.chunk_filenames = Some(format!("[name].{output_ext}"));
      }
    }

    if self.test_meta.visualize_sourcemap {
      if options.sourcemap.is_none() {
        options.sourcemap = Some(SourceMapType::File);
      } else if !matches!(options.sourcemap, Some(SourceMapType::File)) {
        panic!("`visualizeSourcemap` is only supported with `sourcemap: 'file'`")
      }
    }
    if options.sourcemap.is_none() && self.test_meta.visualize_sourcemap {
      options.sourcemap = Some(SourceMapType::File);
    }
  }

  #[allow(clippy::too_many_lines)]
  fn snapshot_bundle_output(&self, bundle_output: BundleOutput, cwd: &Path) {
    let mut snapshot = String::new();
    let mut errors = bundle_output.errors;
    if !errors.is_empty() {
      snapshot.push_str("# Errors\n\n");
      errors.sort_by_key(|e| e.kind().to_string());
      let diagnostics = errors
        .into_iter()
        .map(|e| (e.kind(), e.into_diagnostic_with(&DiagnosticOptions { cwd: cwd.to_path_buf() })));

      let rendered = diagnostics
        .flat_map(|(code, diagnostic)| {
          [
            Cow::Owned(format!("## {code}\n")),
            "```text".into(),
            Cow::Owned(diagnostic.to_string()),
            "```".into(),
          ]
        })
        .collect::<Vec<_>>()
        .join("\n");
      snapshot.push_str(&rendered);
    }

    // Make the snapshot consistent
    let mut warnings = bundle_output.warnings;
    if !warnings.is_empty() {
      warnings.sort_by(|a, b| {
        let a = a.to_string();
        let b = b.to_string();
        a.cmp(&b)
      });
      snapshot.push_str("# warnings\n\n");
      let diagnostics = warnings
        .into_iter()
        .map(|e| (e.kind(), e.into_diagnostic_with(&DiagnosticOptions { cwd: cwd.to_path_buf() })));
      let rendered = diagnostics
        .flat_map(|(code, diagnostic)| {
          [
            Cow::Owned(format!("## {code}\n")),
            "```text".into(),
            Cow::Owned(diagnostic.to_string()),
            "```".into(),
          ]
        })
        .collect::<Vec<_>>()
        .join("\n");
      snapshot.push_str(&rendered);
      snapshot.push('\n');
    }

    let mut assets = bundle_output.assets;

    if !assets.is_empty() {
      snapshot.push_str("# Assets\n\n");
      assets.sort_by_key(|c| c.filename().to_string());
      let artifacts = assets
        .iter()
        .filter(|asset| {
          !asset.filename().contains("$runtime$") && matches!(asset, Output::Chunk(_))
        })
        .flat_map(|asset| {
          let content = std::str::from_utf8(asset.content_as_bytes()).unwrap();
          let content = if self.test_meta.hidden_runtime_module {
            RUNTIME_MODULE_OUTPUT_RE.replace_all(content, "")
          } else {
            Cow::Borrowed(content)
          };

          [Cow::Owned(format!("## {}\n", asset.filename())), "```js".into(), content, "```".into()]
        })
        .collect::<Vec<_>>()
        .join("\n");
      snapshot.push_str(&artifacts);
    }

    if self.test_meta.snapshot_output_stats {
      snapshot.push_str("\n\n## Output Stats\n\n");
      let stats = assets
        .iter()
        .flat_map(|asset| match asset {
          Output::Chunk(chunk) => {
            vec![Cow::Owned(format!(
              "- {}, is_entry {}, is_dynamic_entry {}, exports {:?}",
              chunk.filename.as_str(),
              chunk.is_entry,
              chunk.is_dynamic_entry,
              chunk.exports
            ))]
          }
          Output::Asset(_) => vec![],
        })
        .collect::<Vec<_>>()
        .join("\n");
      snapshot.push_str(&stats);
    }

    if self.test_meta.visualize_sourcemap {
      snapshot.push_str("\n\n# Sourcemap Visualizer\n\n");
      let visualizer_result = assets
        .iter()
        .filter_map(|asset| match asset {
          Output::Chunk(chunk) => chunk.map.as_ref().map(|sourcemap| {
            SourcemapVisualizer::new(&chunk.code, sourcemap).into_visualizer_text()
          }),
          Output::Asset(_) => None,
        })
        .collect::<Vec<_>>()
        .join("\n");
      snapshot.push_str(&visualizer_result);
    }

    // Configure insta to use the fixture path as the snapshot path
    let mut settings = insta::Settings::clone_current();
    settings.set_snapshot_path(cwd);
    settings.set_prepend_module_to_snapshot(false);
    settings.set_input_file(cwd);
    settings.bind(|| {
      insta::assert_snapshot!("artifacts", snapshot);
    });
  }

  fn execute_output_assets(bundler: &Bundler) {
    let mut command = Command::new("node");

    let cwd = bundler.options().cwd.clone();
    let dist_folder = cwd.join(&bundler.options().dir);
    let is_output_cjs = matches!(bundler.options().format, OutputFormat::Cjs);

    let test_script = if is_output_cjs { cwd.join("_test.cjs") } else { cwd.join("_test.mjs") };

    let compiled_entries = bundler
      .options()
      .input
      .iter()
      .map(|item| {
        let name = item.name.clone().expect("inputs must have `name` in `_config.json`");
        let ext = if is_output_cjs { "cjs" } else { "mjs" };
        format!("{name}.{ext}",)
      })
      .map(|name| dist_folder.join(name))
      .collect::<Vec<_>>();

    compiled_entries.iter().for_each(|entry| {
      if is_output_cjs {
        command.arg("--require");
      } else {
        command.arg("--import");
      }
      if cfg!(target_os = "windows") && !is_output_cjs {
        // Only URLs with a scheme in: file, data, and node are supported by the default ESM loader. On Windows, absolute paths must be valid file:// URLs.
        command.arg(format!("file://{}", entry.to_str().expect("should be valid utf8")));
      } else {
        command.arg(entry);
      }
    });

    if test_script.exists() {
      command.arg(test_script);
    } else {
      command.arg("--eval");
      command.arg("\"\"");
    }

    let output = command.output().unwrap();

    #[allow(clippy::print_stdout)]
    if !output.status.success() {
      let stdout_utf8 = std::str::from_utf8(&output.stdout).unwrap();
      let stderr_utf8 = std::str::from_utf8(&output.stderr).unwrap();

      println!("⬇️⬇️ Failed to execute command ⬇️⬇️\n{command:?}\n⬆️⬆️ end  ⬆️⬆️");
      panic!("⬇️⬇️ stderr ⬇️⬇️\n{stderr_utf8}\n⬇️⬇️ stdout ⬇️⬇️\n{stdout_utf8}\n⬆️⬆️ end  ⬆️⬆️",);
    }
  }
}
