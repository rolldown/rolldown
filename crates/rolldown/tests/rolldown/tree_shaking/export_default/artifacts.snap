---
source: crates/rolldown_testing/src/integration_test.rs
---
# Assets

## main.mjs

```js

//#region foo.js
var foo_default = 1;

//#endregion
//#region export.js
var export_default = { foo: foo_default };

//#endregion
//#region main.js
export_default.foo;

//#endregion
```
